File: integration_test/tests/super_class.rs
use integration_test::run_class;

// #[futures_test::test]
async fn test_superclass() -> anyhow::Result<()> {
    let super_class = include_bytes!("../../test_data/SuperClass.class");
    let inner = include_bytes!("../../test_data/SuperClass$InnerClass.class");
    let inner_derived = include_bytes!("../../test_data/SuperClass$InnerDerivedClass.class");

    let result = run_class(
        "SuperClass",
        &[
            ("SuperClass", super_class),
            ("SuperClass$InnerClass", inner),
            ("SuperClass$InnerDerivedClass", inner_derived),
        ],
        &[],
    )
    .await?;
    assert_eq!(result, "2\n1234123412341234\n2\ntest\ntest\n");

    Ok(())
}

File: test_data/src/SuperClass.java
class SuperClass {
    class InnerClass {
        int field1;
        long field2;

        void test() {
            System.out.println("test");
        }
    }

    class InnerDerivedClass extends InnerClass {
        int field3;
        String field4;
    }

    void run() {
        InnerDerivedClass derived = null;

        derived = new InnerDerivedClass();

        derived.field1 = 0;
        derived.field2 = 1;
        derived.field3 = 2;
        derived.field4 = "test";

        InnerClass inner = derived;

        inner.field1 = 2;
        inner.field2 = 1234123412341234L;

        System.out.println(derived.field1);
        System.out.println(derived.field2);
        System.out.println(derived.field3);
        System.out.println(derived.field4);

        derived.test();
    }

    public static void main(String[] args) {
        new SuperClass().run();
    }
}
